site/src/modules/apps/apps.ts

```diff
@@ -49,6 +61,73 @@ export const getTerminalHref = ({
 	}/terminal?${params}`;
 };
 
-export const openAppInNewWindow = (name: string, href: string) => {
+export const openAppInNewWindow = (href: string) => {
 	window.open(href, "_blank", "width=900,height=600");
 };
+
+export type GetAppHrefParams = {
+	path: string;
+	host: string;
+	workspace: Workspace;
+	agent: WorkspaceAgent;
+	token?: string;
+};
+
+export const getAppHref = (
+	app: WorkspaceApp,
+	{ path, token, workspace, agent, host }: GetAppHrefParams,
+): string => {
+	if (isExternalApp(app)) {
+		return needsSessionToken(app)
+			? app.url.replaceAll(SESSION_TOKEN_PLACEHOLDER, token ?? "")
+			: app.url;
+	}
+
+	// The backend redirects if the trailing slash isn't included, so we add it
+	// here to avoid extra roundtrips.
+	let href = `${path}/@${workspace.owner_name}/${workspace.name}.${
+		agent.name
+	}/apps/${encodeURIComponent(app.slug)}/`;
+
+	if (app.command) {
+		// Terminal links are relative. The terminal page knows how
+		// to select the correct workspace proxy for the websocket
+		// connection.
+		href = `/@${workspace.owner_name}/${workspace.name}.${
+			agent.name
+		}/terminal?command=${encodeURIComponent(app.command)}`;
+	}
+
+	if (host && app.subdomain && app.subdomain_name) {
+		const baseUrl = `${window.location.protocol}//${host.replace(/\*/g, app.subdomain_name)}`;
+		const url = new URL(baseUrl);
+		url.pathname = "/";
+		href = url.toString();
+	}
+
+	return href;
```

Comment:
I know this was already like this, so no need to change, but it does seem like early returns could make more sense than a `let`, since we are not ever combining anything, just returning three separate hrefs.

Reply:
Going to refactor this in a next PR.